src/main/webapp/app/core/util/alert.service.ts

```diff
@@ -1,38 +1,114 @@
-import { Injectable, SecurityContext, NgZone } from '@angular/core';
+import { Injectable, NgZone, SecurityContext } from '@angular/core';
 import { DomSanitizer } from '@angular/platform-browser';
 import { TranslateService } from '@ngx-translate/core';
 import { translationNotFoundMessage } from 'app/core/config/translation.config';
+import { EventManager, EventWithContent } from 'app/core/util/event-manager.service';
+import { AlertError } from 'app/shared/alert/alert-error.model';
+import { Subscription } from 'rxjs';
 
 export type AlertType = 'success' | 'danger' | 'warning' | 'info';
 
-export interface Alert {
-    id?: number;
+export interface AlertBase {
     type: AlertType;
     message?: string;
     translationKey?: string;
     translationParams?: { [key: string]: unknown };
     timeout?: number;
-    toast?: boolean;
-    position?: string;
-    close?: (alerts: Alert[]) => void;
+    action?: { label: string; callback: (alert: Alert) => void };
+    onClose?: (alert: Alert) => void;
+    dismissible?: boolean;
+}
+
+export interface Alert extends AlertBase {
+    id: number;
+    close: () => void;
+    readonly isOpen: boolean;
+}
+
+export interface AlertInternal extends Alert {
+    closeFunction?: (callback: () => void) => void;
+    isOpen: boolean;
 }
 
 @Injectable({
     providedIn: 'root',
 })
 export class AlertService {
     timeout = 8000;
-    toast = false;
-    position = 'top right';
+    dismissible = true;
 
     // unique id for each alert. Starts from 0.
     private alertId = 0;
-    private alerts: Alert[] = [];
+    private alerts: AlertInternal[] = [];
 
-    constructor(private sanitizer: DomSanitizer, private ngZone: NgZone, private translateService: TranslateService) {}
+    errorListener: Subscription;
+    httpErrorListener: Subscription;
+
+    constructor(private sanitizer: DomSanitizer, private ngZone: NgZone, private translateService: TranslateService, private eventManager: EventManager) {
+        this.errorListener = eventManager.subscribe('artemisApp.error', (response: EventWithContent<unknown> | string) => {
+            const errorResponse = (response as EventWithContent<AlertError>).content;
+            this.addErrorAlert(errorResponse.message, errorResponse.translationKey, errorResponse.translationParams);
+        });
+
+        this.httpErrorListener = eventManager.subscribe('artemisApp.httpError', (response: any) => {
+            const httpErrorResponse = response.content;
```

Comment:
This block is quite long with a lot of indentation. Can you maybe break it up into smaller methods?

Reply:
Possibly. However, I think it's out of scope for this PR. I want to rework how the alerts look, feel and work, rather than changing the code to catch http errors. This is copied from the old implementation and has not been touched